dd-sdk-android-core/src/main/kotlin/com/datadog/android/core/internal/data/upload/v2/DataUploadRunnable.kt

```diff
@@ -54,28 +55,40 @@ internal class DataUploadRunnable(
             val context = contextProvider.context
             // TODO RUMM-0000 it should be already on the worker thread and if readNextBatch is async,
             //  we should wait until it completes before scheduling further
-            val lock = CountDownLatch(1)
-            storage.readNextBatch(
-                noBatchCallback = {
-                    increaseInterval()
-                    lock.countDown()
+            var batchesSent = 0
+            var lastUploadWasSuccess = false
+            do  {
+                val lock = CountDownLatch(1)
+                storage.readNextBatch(
+                    noBatchCallback = {
+                        increaseInterval()
+                        lock.countDown()
+                        batchesSent = maxBatchesPersUpload
+                        lastUploadWasSuccess = false
+                    }
+                ) { batchId, reader ->
+                    try {
+                        val batch = reader.read()
+                        val batchMeta = reader.currentMetadata()
+
+                        val status = consumeBatch(
+                            context,
+                            batchId,
+                            batch,
+                            batchMeta
+                        )
+                        if (status !is UploadStatus.Success) {
+                            batchesSent = maxBatchesPersUpload
```

Comment:
this is not quite true, why relying only on `lastUploadWasSuccess` is not enough?

Reply:
yes was a leftover, sorry for the trouble ;)